### PR TITLE
Prompt Explorer

### PR DESCRIPTION
**Pull request by dgtm777**
Tool for data and prompt exploration.
The tool has two pages:
- Run prompt: Allows to explore prompt and interact with running model
- Analyze: Allows to explore generated data

### CODE REVIEW

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:1]:
```suggestion
Before using this tool, you first need to download datasets and install requirements:
```

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:7]: Let's not have a default here and require to explicitly set it instead

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:11]:
```suggestion
    python3 visualization/main.py \
```
or maybe rename it to data_explorer.py to make it more verbose

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:18]: this is not necessary

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:27]: Can it be changed from command line?

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:30]: There is no context as to what this means, can you explain why this is needed and how it's going to be used? You can also add some images showing how the tool works, similar to how it'd done for the speech data explorer docs in nemo

File: visualization/callbacks/__init__.py
Comment (Kipok) [inline visualization/callbacks/__init__.py:1]: Please add license header to all files

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:14]: Some details are missing - e.g. do I need to launch server myself before starting this script?

File: visualization/callbacks/__init__.py
Comment (Kipok) [inline visualization/callbacks/__init__.py:21]: Let's not call it "prompt_explorer", but "data_explorer", since it's not just for prompt, right?

File: visualization/main.py
Comment (Kipok) [inline visualization/main.py:21]: How does it get the arguments?

File: visualization/settings/config.py
Comment (Kipok) [inline visualization/settings/config.py:12]: why is this a dict? Please use different name if it's not the same as prediction_jsonl_files in other configs

File: visualization/settings/config.py
Comment (Kipok) [inline visualization/settings/config.py:16]: the default is confusing here, what do the brackets mean? Why don't you take dataset and split_name is parameters like in other files?

File: visualization/settings/templates.py
Comment (Kipok) [inline visualization/settings/templates.py:10]:
```suggestion
python nemo_skills/inference/generate_solutions.py \\
```
or use sys.executable like here https://github.com/Kipok/NeMo-Skills/blob/main/pipeline/summarize_results.py#L55

Comment (Kipok) [inline visualization/settings/templates.py:10]: Wouldn't it be easier to directly call main from generate_solutions and pass in config file? Then you don't need to map it like this if you have it as a config object in the code

File: visualization/settings/templates.py
Comment (Kipok) [inline visualization/settings/templates.py:5]: I think you need to package installed anyway, so maybe remove python path and just ask to follow instructions in prerequisites docs for local installation?

File: visualization/settings/templates.py
Comment (Kipok) [inline visualization/settings/templates.py:4]:
```suggestion
path = str(Path(__file__).parents[2].absolute())
```

File: visualization/settings/visualization_config.yaml
Comment (Kipok) [inline visualization/settings/visualization_config.yaml:3]:
```suggestion
    model1: ???
```

File: visualization/utils/common.py
Comment (Kipok) [inline visualization/utils/common.py:54]: there are functions in nemo_skills/code_execution/utils.py and nemo_skills/code_execution/math_grader.py that can extract code / errors / answers that you can reuse?

File: visualization/utils/common.py
Comment (Kipok) [inline visualization/utils/common.py:108]: please split into multiple expressions, so that black does not autoformat this like that

File: visualization/utils/common.py
Comment (Kipok) [inline visualization/utils/common.py:241]: why is this needed?

File: visualization/utils/strategies/base_strategy.py
Comment (Kipok) [inline visualization/utils/strategies/base_strategy.py:27]: why don't you reuse the config files information here? The sandbox does not have to be on the same host + in the future we'd have more sandboxes with different parameters, so better to have generic init here

File: visualization/utils/strategies/base_strategy.py
Comment (Kipok) [inline visualization/utils/strategies/base_strategy.py:175]: same here, should come directly from the config so that it's possible to customize other potential parameters

File: visualization/utils/strategies/whole_dataset_mode.py
Comment (Kipok) [inline visualization/utils/strategies/whole_dataset_mode.py:16]: Is this a mistake? Should it be WHOLE_DATASET_MODE?

**Review by Kipok — CHANGES_REQUESTED** [review]
Thanks @dgtm777, great work!

Overall, the biggest missing part is the lack of some documentation explaining what can be done with the tool (can add images as well to illustrate). Also would be great (but can be in follow up PR) to add some "development" docs, so that you can at least give a basic explanation of how to implement new features as the code looks quite involved.

File: visualization/Readme.md
Comment (dgtm777) [reply visualization/Readme.md:27]: Path to files should be a Dict[str, List[str]] (Dict keys are model names and values are a list of files) it is tricky to have it in command line

File: visualization/main.py
Comment (dgtm777) [reply visualization/main.py:21]: Arguments are in the app variable

File: visualization/utils/common.py
Comment (dgtm777) [reply visualization/utils/common.py:241]: 1. it works faster
2. it does not do deepcopy for items in last json. So when I update them they are updated everywhere

File: .gitignore
Comment (Kipok) [inline .gitignore:26]: this are a bit too general names, maybe put them inside visualization folder by default and ignore there?

File: nemo_skills/inference/prompt/utils.py
Comment (Kipok) [inline nemo_skills/inference/prompt/utils.py:45]: This is being refactored in https://github.com/Kipok/NeMo-Skills/pull/10, so maybe we should switch to that version before merging this. @i-vainn please sync with Daria to make sure the new format supports her use-cases and let's prioritize merging your PR to unblock this one

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:4]: Should this be split into multiple lines? You can use some tool (e.g. vs-code) to preview the markdown, right now this is all just one really big line

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:2]:
```suggestion
## Run Prompt page
```

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:1]: run model or run prompt? Pick one

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:6]: Let's just save metrics/outputs to one folder, so that you only need to specify folder name?

Comment (Kipok) [reply visualization/Readme.md:6]: Should it be `end_random_seed`?

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:8]:
```suggestion
To use the Analyze page, you need to specify paths to the datasets you want to analyze (if you did not get the data through the "run model" page). You can pass parameters via command line  `++visualization_params.model_prediction.model1='/some_path/model1/output-greedy.jsonl'` or modify `visualization_config.py`.
```
Maybe name it like "inference" / "analyze" ?

Comment (Kipok) [reply visualization/Readme.md:8]: visualization_config.py -> visualization_config.yaml (better add a link)

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:17]: Put this on top

File: visualization/callbacks/__init__.py
Comment (Kipok) [inline visualization/callbacks/__init__.py:47]: This is a bit confusing - can you add some comments on why you need to pop all of this?

File: visualization/settings/config.py
Comment (Kipok) [inline visualization/settings/config.py:27]: why do you need a default to empty dict? Shouldn't it be MISSING instead?

File: visualization/settings/visualization_config.yaml
Comment (Kipok) [inline visualization/settings/visualization_config.yaml:3]: one model is also supported, right? Should change defaults then? Also what about 3+ models? If not supported, better to add some checks for that, I guess

File: visualization/utils/common.py
Comment (Kipok) [inline visualization/utils/common.py:36]: Can you just import all examples like done here https://github.com/Kipok/NeMo-Skills/blob/main/nemo_skills/inference/prompt/utils.py#L27

File: visualization/utils/common.py
Comment (Kipok) [reply visualization/utils/common.py:68]: I don't quite understand - can you please explain this more?

File: visualization/utils/strategies/whole_dataset_mode.py
Comment (Kipok) [inline visualization/utils/strategies/whole_dataset_mode.py:108]: Can you do the same for evaluation? Call a function instead of running script?

**Review by Kipok — CHANGES_REQUESTED** [review]
Thanks, looks better, but needs a bit more work!

File: visualization/settings/config.py
Comment (dgtm777) [reply visualization/settings/config.py:27]: The "run model" page doesn't need this parameter, so I don't want to force it to be filled in

File: visualization/utils/strategies/whole_dataset_mode.py
Comment (dgtm777) [reply visualization/utils/strategies/whole_dataset_mode.py:108]: There is a problem with typings. prediction_jsonl_files should be a str when I do OmegaConf.structured, but in __post_init__ it becomes List[str] and generate_solutions expect it to be List[str]

File: visualization/utils/common.py
Comment (dgtm777) [reply visualization/utils/common.py:68]: I have the full text as an input and I should extract text, code, and code_output blocks preserving their order. 
extract_error_message from nemo_skills/code_execution/utils.py extracts only last code output block
extract_code_to_execute from nemo_skills/code_execution/utils.py extracts only last code block
extract_answer from nemo_skills/code_execution/math_grader.py extraxtc only value insibe \boxed{}

File: visualization/settings/visualization_config.yaml
Comment (Kipok) [reply visualization/settings/visualization_config.yaml:3]: then better to not have this filled in to two models by default? Maybe just keep it empty or remove this config completely (you already have a dataclass, right)?

File: visualization/utils/strategies/whole_dataset_mode.py
Comment (Kipok) [reply visualization/utils/strategies/whole_dataset_mode.py:109]: can you set it to Any then if it fails on the type? Or maybe there is some parameter in OmegaConf that you can use to ignore types?

File: visualization/callbacks/__init__.py
Comment (Kipok) [inline visualization/callbacks/__init__.py:22]: Need to remove?

**Review by Kipok — CHANGES_REQUESTED** [review]

File: nemo_skills/inference/prompt/utils.py
Comment (Kipok) [reply nemo_skills/inference/prompt/utils.py:45]: The PR is merged, please integrate it into your branch

File: .gitignore
Comment (Kipok) [inline .gitignore:27]: visualization/ ?

File: nemo_skills/evaluation/evaluate_results.py
Comment (Kipok) [inline nemo_skills/evaluation/evaluate_results.py:21]: Is this used?

File: nemo_skills/inference/generate_solutions.py
Comment (Kipok) [inline nemo_skills/inference/generate_solutions.py:56]: why comment removed?

File: nemo_skills/inference/generate_solutions.py
Comment (Kipok) [inline nemo_skills/inference/generate_solutions.py:65]: Please add docstring in a comment

File: nemo_skills/inference/prompt/text_masked_base.yaml
Comment (Kipok) [inline nemo_skills/inference/prompt/text_masked_base.yaml:11]: "masked" missing here and on top

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:7]: This is already part of prerequisites, so can skip listing it here

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:11]: I think the main use-case is actually running this locally, not on slurm - so maybe list just the host as an argument and mention that can also add ssh parameters if this is necessary for connecting to slurm

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:19]:
```suggestion
This page enables the analysis of model answers based on different parameters. It offers three modes: "Chat", "Run one sample", and "Run whole dataset".
```

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:23]:
```suggestion
- **Run whole dataset** mode lets you launch the model with chosen parameters on the entire dataset. Results are saved in `visualization/results/output-greedy.jsonl` and `visualization/results/metrics-greedy.jsonl`. If the "use random seed range" flag is enabled, each answer will be sampled with multiple random seeds in the range from `start_random_seed` to `end_random_seed`. After generation is done, you can review the results on the "Analyze" page. The parameters used for dataset launch are also recorded in the `visualization/results/parameters.jsonl` file and displayed on the "Analyze" page.
```

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:35]:
```suggestion
The tool also supports comparison of multiple model outputs (e.g. 
 `model1` and `model2` in the config above). All files satisfying the given pattern will be considered for analysis.
```

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:37]:
```suggestion
On this page, you can sort, filter, and compare model outputs. You can also add labels to the data and save your modified, filtered, and sorted dataset by specifying `save_dataset_path`.
```

**Review by Kipok — CHANGES_REQUESTED** [review]

File: visualization/Readme.md
Comment (Kipok) [inline visualization/Readme.md:13]:
```suggestion
For the "Inference" page, launch the server with the model (see [inference.md](/docs/inference.md)), specify `host` and, if necessary, `ssh_key` and `ssh_server`.
```

**Review by Kipok — APPROVED** [review]
Looks good, thanks for the great work @dgtm777! Let's just add video tutorials and we are good to merge.

File: pipeline/summarize_results.py
Comment (Kipok) [inline pipeline/summarize_results.py:52]:
```suggestion
        for benchmark in benchmarks.copy():
```

File: pipeline/start_server.py
Comment (Kipok) [inline pipeline/start_server.py:39]: Please double check that it works locally

**Review by Kipok — APPROVED** [review]

**Pull request merged**
